File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/NQueens/NQueens_CSharp/MainForm.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: MainForm.cs
//
//--------------------------------------------------------------------------

using System;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.Windows.Forms.DataVisualization.Charting;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples;

namespace NQueens_CSharp
{
    public partial class MainForm : Form
    {
        const int NUM_QUEENS = 22;
        private TaskScheduler _uiScheduler;
        private DataPoint _serial;
        private DataPoint _parallel;
        private double _max = 0;

        public MainForm()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            _uiScheduler = TaskScheduler.FromCurrentSynchronizationContext();
            SetupPoints();
        }

        private void SetupPoints()
        {
            _serial = chart1.Series[0].Points[chart1.Series[0].Points.AddXY("Serial", 0.0)];
            _parallel = chart1.Series[0].Points[chart1.Series[0].Points.AddXY("Parallel", 0.0)];
            chart1.ChartAreas[0].AxisY.Minimum = 0.0;
            chart1.ChartAreas[0].AxisY.Maximum = 10.0;
            ClearPointValues();
            Invalidate();
        }

        private void ClearPointValues()
        {
            foreach (var point in new[] { _serial, _parallel })
            {
                point.SetValueY(0);
                point.ToolTip = "";
                point.Label = "";
                point.Font = new System.Drawing.Font(point.Font, System.Drawing.FontStyle.Bold);
            }
            chart1.Invalidate();
        }

        private void btnSolve_Click(object sender, EventArgs e)
        {
            int numQueens = 0;
            if (!Int32.TryParse(cbNumQueens.Text, out numQueens)) return;

            btnSolve.Enabled = false;
            progressBar1.Visible = true;

            ClearPointValues();
            _max = 0;

            Task.Factory.StartNew(() =>
            {
                TimeSpan elapsed;

                elapsed = Time(() => NQueensSolver.Sequential(numQueens));
                SetPoint(_serial, elapsed.TotalSeconds);

                elapsed = Time(() => NQueensSolver.Parallel(numQueens));
                SetPoint(_parallel, elapsed.TotalSeconds);

            }, TaskCreationOptions.AttachedToParent).ContinueWith(t =>
            {
                progressBar1.Visible = false;
                btnSolve.Enabled = true;
                if (t.IsFaulted) MessageBox.Show(t.Exception.ToString());
            }, CancellationToken.None, TaskContinuationOptions.AttachedToParent, _uiScheduler);
        }

        private void SetPoint(DataPoint point, double seconds)
        {
            Task.Factory.StartNew(() =>
            {
                if (seconds > _max)
                {
                    _max = seconds;
                    chart1.ChartAreas[0].AxisY.Maximum = _max * 1.1;
                }
                point.SetValueY(seconds);
                point.ToolTip = seconds.ToString("F3");
                point.Label = string.Format("{0:F3} secs  ({1:F2}x)", seconds, (_max / seconds));
                chart1.Invalidate();
            }, CancellationToken.None, TaskCreationOptions.AttachedToParent, _uiScheduler);
        }

        static TimeSpan Time(Action action)
        {
            var sw = Stopwatch.StartNew();
            action();
            return sw.Elapsed;
        }
    }
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/NQueens/NQueens_VisualBasic/MainForm.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: MainForm.vb
'
'--------------------------------------------------------------------------

Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms.DataVisualization.Charting

Public Class MainForm

    Private _uiScheduler As TaskScheduler
    Private _serial As DataPoint
    Private _parallel As DataPoint
    Private _max As Double = 0

    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        _uiScheduler = TaskScheduler.FromCurrentSynchronizationContext()
        SetupPoints()
    End Sub

    Private Sub SetupPoints()
        _serial = chart1.Series(0).Points(chart1.Series(0).Points.AddXY("Serial", 0.0))
        _parallel = chart1.Series(0).Points(chart1.Series(0).Points.AddXY("Parallel", 0.0))
        chart1.ChartAreas(0).AxisY.Minimum = 0.0
        chart1.ChartAreas(0).AxisY.Maximum = 10.0
        ClearPointValues()
        Invalidate()
    End Sub

    Private Sub ClearPointValues()
        For Each Point In New DataPoint() {_serial, _parallel}
            With Point
                .SetValueY(0)
                .ToolTip = ""
                .Label = ""
                .Font = New System.Drawing.Font(Point.Font, System.Drawing.FontStyle.Bold)
            End With

        Next
        chart1.Invalidate()
    End Sub


    Private Sub btnSolve_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSolve.Click
        Dim numQueens = 0
        If Int32.TryParse(cbNumQueens.Text, numQueens) = False Then Return

        btnSolve.Enabled = False
        progressBar1.Visible = True

        ClearPointValues()
        _max = 0

        Task.Factory.StartNew(Sub()
                                  Dim elapsed As TimeSpan

                                  elapsed = Time(Sub()
                                                     NQueensSolver.Sequential(numQueens)
                                                 End Sub)
                                  SetPoint(_serial, elapsed.TotalSeconds)

                                  elapsed = Time(Sub()
                                                     NQueensSolver.Parallel(numQueens)
                                                 End Sub)
                                  SetPoint(_parallel, elapsed.TotalSeconds)
                              End Sub
        ).ContinueWith(Sub(t)
                           progressBar1.Visible = False
                           btnSolve.Enabled = True
                           If t.IsFaulted Then  MessageBox.Show(t.Exception.ToString())
                       End Sub, _uiScheduler)
    End Sub


    Private Sub SetPoint(ByVal point As DataPoint, ByVal seconds As Double)
        Task.Factory.StartNew(Sub()
                                  If seconds > _max Then
                                      _max = seconds
                                      chart1.ChartAreas(0).AxisY.Maximum = _max * 1.1
                                  End If
                                  point.SetValueY(seconds)
                                  point.ToolTip = seconds.ToString("F3")
                                  point.Label = String.Format("{0:F3} secs  ({1:F2}x)", seconds, (_max / seconds))
                                  chart1.Invalidate()
                              End Sub, CancellationToken.None, TaskCreationOptions.None, _uiScheduler)
    End Sub

    Private Shared Function Time(ByVal action As Action) As TimeSpan
        Dim sw = Stopwatch.StartNew()
        action()
        Return sw.Elapsed
    End Function
End Class
